KSP-122 Refactor UserContext

navarrojandg opened the pull request:
Spinner should not be used in the UserAuthProvider.

UserAuthProvider is a higher level component that wraps its children and provides the currentUser state, placing a spinner here does not make sense because its meant do delegate state and operates in the background so it shouldnt be responsible for conveying any information to the user.

ivancuadra869 commented:
> Spinner should not be used in the UserAuthProvider.
> 
> UserAuthProvider is a higher level component that wraps its children and provides the currentUser state, placing a spinner here does not make sense because its meant do delegate state and operates in the background so it shouldnt be responsible for conveying any information to the user.

So this wasn't doing anything before huh, from the looks of it.

navarrojandg commented:
Well also more importantly if that if pending block executed it would break
the rest of the app bc now all child components wanting to useUserContext
cannot since it returns a spinner rather than the context provider.

On Sun, Nov 14, 2021 at 19:36 ivancuadra869 ***@***.***>
wrote:

> Spinner should not be used in the UserAuthProvider.
>
> UserAuthProvider is a higher level component that wraps its children and
> provides the currentUser state, placing a spinner here does not make sense
> because its meant do delegate state and operates in the background so it
> shouldnt be responsible for conveying any information to the user.
>
> So this wasn't doing anything before huh, from the looks of it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CSUN-COMP-490-Team-KBBQ/wya-app/pull/43#issuecomment-968507808>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKCKTNC7CYPAOHUWJFDF5ODUMB55FANCNFSM5IAVUBWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jesc0bar commented:
Pending is a third state which is describing the loading for a user auth.

jesc0bar commented:
> Well also more importantly if that if pending block executed it would break the rest of the app bc now all child components wanting to useUserContext cannot since it returns a spinner rather than the context provider.
> […](#)
> On Sun, Nov 14, 2021 at 19:36 ivancuadra869 ***@***.***> wrote: Spinner should not be used in the UserAuthProvider. UserAuthProvider is a higher level component that wraps its children and provides the currentUser state, placing a spinner here does not make sense because its meant do delegate state and operates in the background so it shouldnt be responsible for conveying any information to the user. So this wasn't doing anything before huh, from the looks of it. — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <[#43 (comment)](https://github.com/CSUN-COMP-490-Team-KBBQ/wya-app/pull/43#issuecomment-968507808)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKCKTNC7CYPAOHUWJFDF5ODUMB55FANCNFSM5IAVUBWA> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

The pending state should not give any information to the user since the auth is yet to be verified.

jesc0bar commented:
> Spinner should not be used in the UserAuthProvider.
> 
> UserAuthProvider is a higher level component that wraps its children and provides the currentUser state, placing a spinner here does not make sense because its meant do delegate state and operates in the background so it shouldnt be responsible for conveying any information to the user.

I agree that there should not be a render here. We could make a simple component instead but pending is to signify the third state which is loading.

ivancuadra869 commented on src/contexts/UserRecordContext.tsx:
```diff
@@ -3,21 +3,37 @@ import UserData from '../interfaces/User';
+                    }
+                },
+            });
+        }
```
So while testing I noticed that clicking on "Log out" does not reflect properly. The "Log out" is only apparent after refreshing the page.

Jorge and I made some changes and noticed that the following seemed to fix it

```    
React.useEffect(() => {
        if (!pending && user) {
            return getDocSnapshot$(`/users/${user.uid}`, {
                next: (snapshot) => {
                    if (snapshot.exists()) {
                        setUserRecord({
                            pending: false,
                            userRecord: snapshot.data() as UserData,
                        });
                    }
                },
            });
        }
        setUserRecord({
            pending: pending,
            userRecord: null,
        });
    }, [user]); 
```

navarrojandg replied:
Yeah good catch!

navarrojandg commented:
Fixed conflict with main's RegisterForm.

Added back the consistent return eslint rule because this warning came up on UserRecordProvider's useEffect hook and should have been a sign that we were not updating states for all possibilities.

ivancuadra869 reviewed (approved):
Looks good!

jesc0bar commented on src/components/LoginForm/LoginForm.tsx:
```diff
@@ -11,9 +11,9 @@ import './LoginForm.css';
+    const { pending, user } = useUserContext();
     React.useEffect(() => {
-        if (user) {
+        if (!pending && user) {
```
What effect does `pending` have on this component?

navarrojandg replied:
What I was thinking is that when we go from `{pending: false, user: null}` -> `{pending:true, user: null}` after the user enters their email and password to log in rather that just flashing to a logged in state we can show a spinner here when user credentials are being fetched.

jesc0bar commented on src/components/RegisterForm/RegisterForm.tsx:
```diff
@@ -14,7 +14,7 @@ import './RegisterForm.css';
 export default function RegisterForm(): JSX.Element {
     const history = useHistory();
-    const user = useUserContext();
+    const { user } = useUserContext();
```
`pending` was used for login form. Does it need to be use here too? 

Although, I'm really clear on if `pending` affects the outcome.

navarrojandg replied:
Mmm yeah we can use `pending` here similar to log in because after a new user goes to create an account we can show a spinner as well.

jesc0bar reviewed (approved):
Looks good.

I do have concerns on the complexity of the states and how there isn't a central location for the loading component. Right now there isn't a uniform flow for showing a loading stage and handling states with this level of complexity. This is largely due to the separation of private and public pages since they are dependent on authorization. An example of a side effect is seen when the home page is reloaded. 

Overall, the core functionality continues to work and the concerns raised could be addressed in a future refactor when there are less pressing tasks. This is why I am approving the PR.

navarrojandg commented:
@jesc0bar I have refactored UserRecordContext with guards against the UserContext. I suggest a similar pattern to be followed when a component / page depends on [UserRecord | User]Context. I've refactored PrivateRoute and HomePage as examples. Other components / pages may need to be refactored in the future but did not want to include those changes in this PR as this PR is getting relatively big.
